File: src/tavern/scene-user-actions.test.ts
import { describe, expect, it } from 'vitest';
import fixture from '../../shared/tavern-scene-fixtures.json';
import { readSceneBindings, applyScenePatch, canonicalSceneValue } from './scene-bindings';
import { sceneEntityRemoval } from './scene-user-actions';
import { sceneErrorMessage } from './scene-errors';

const base = () => readSceneBindings(fixture.scene)!;
describe('review fixes: deliberate scene edits', () => {
  it('removes clothing and its facts atomically without changing either character', () => {
    const scene = base(); const snapshot = canonicalSceneValue(scene);
    const {next, patch} = sceneEntityRemoval(scene, 'coat_a');
    expect(next.entities.some(e => e.id === 'coat_a')).toBe(false);
    expect(next.facts.some(f => f.entityId === 'coat_a')).toBe(false);
    expect(next.entities.filter(e => e.kind === 'character')).toEqual(scene.entities.filter(e => e.kind === 'character'));
    expect(next.revision).toBe(scene.revision + 1);
    expect(canonicalSceneValue(scene)).toBe(snapshot);
    expect(applyScenePatch(scene, patch, true)).toEqual(next);
  });
  it('keeps other people and objects when deleting a person, and detaches only their bindings', () => {
    const scene = base();
    scene.entities.find(e => e.id === 'coat_a')!.wearerId = 'b';
    const {next} = sceneEntityRemoval(scene, 'a');
    expect(readSceneBindings(next)).toEqual(next);
    expect(next.entities.find(e => e.id === 'coat_a')).toMatchObject({wearerId:'b'});
    expect(next.entities.find(e => e.id === 'coat_a')?.ownerId).toBeUndefined();
    expect(next.entities.find(e => e.id === 'b')).toEqual(scene.entities.find(e => e.id === 'b'));
    expect(next.relations.some(r => r.actorId === 'a' || r.targetId === 'a')).toBe(false);
    expect(next.facts.filter(f => f.entityId !== 'a')).toEqual(scene.facts.filter(f => f.entityId !== 'a'));
  });
  it('does not allow stale preview, last character deletion, or deletion of fixed descendants', () => {
    const scene = base(); const plan = sceneEntityRemoval(scene,'a');
    expect(() => applyScenePatch({...scene,revision:1}, plan.patch, true)).toThrow('SCENE_STALE');
    expect(() => sceneEntityRemoval(plan.next,'b')).toThrow('SCENE_LAST_CHARACTER');
    scene.entities.find(e => e.id === 'coat_a')!.locked = true;
    expect(() => sceneEntityRemoval(scene,'a')).toThrow('SCENE_LOCKED');
    expect(() => sceneEntityRemoval(scene,'coat_a')).toThrow('SCENE_LOCKED');
  });
  it('supports large user-confirmed cascades without expanding the AI operation limit', () => {
    const scene=base();
    scene.facts=Array.from({length:80},(_,i)=>({id:`f_${i}`,entityId:'coat_a',slot:`s${i}`,prompt:'red'}));
    const {next,patch}=sceneEntityRemoval(scene,'coat_a');
    expect(next.facts).toHaveLength(0);
    expect(()=>applyScenePatch(scene,patch)).toThrow('SCENE_STALE');
  });
  for (const language of ['zh-CN','zh-TW','en-US','ja-JP','ko-KR']) {
    it(`explains all scene codes in ${language} without leaking raw enums`, () => {
      for (const code of ['STALE','LOCKED','REBIND','OPERATION','INVALID','MODEL_CAPACITY','LAST_CHARACTER']) {
        const message=sceneErrorMessage(new Error(`prefix SCENE_${code}`),language);
        expect(message.length).toBeGreaterThan(12); expect(message).not.toContain('SCENE_');
      }
      expect(sceneErrorMessage('Connection timed out',language)).toBe('Connection timed out');
    });
  }
});

File: mobile/test/scene_user_actions_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:novelai_mobile/agent/scene_bindings.dart';
import 'package:novelai_mobile/agent/scene_user_actions.dart';

Map<String, dynamic> base() => jsonDecode(
    File('../shared/tavern-scene-fixtures.json').readAsStringSync())['scene'];
void main() {
  test('remove garment is atomic; other character and garments unchanged', () {
    final scene = base(), before = canonicalSceneValue(base());
    final next = sceneEntityRemoval(scene, 'coat_a')['next'];
    expect(next['entities'].where((e) => e['id'] == 'coat_a'), isEmpty);
    expect(next['facts'].where((e) => e['entityId'] == 'coat_a'), isEmpty);
    expect(next['entities'].where((e) => e['kind'] == 'character'),
        scene['entities'].where((e) => e['kind'] == 'character'));
    expect(canonicalSceneValue(scene), before);
  });
  test('person removal detaches belongings and preserves another wearer', () {
    final scene = base();
    scene['entities'].firstWhere((e) => e['id'] == 'coat_a')['wearerId'] = 'b';
    final next = sceneEntityRemoval(scene, 'a')['next'];
    final coat = next['entities'].firstWhere((e) => e['id'] == 'coat_a');
    expect(coat['wearerId'], 'b');
    expect(coat.containsKey('ownerId'), false);
    expect(
        next['relations']
            .where((r) => r['actorId'] == 'a' || r['targetId'] == 'a'),
        isEmpty);
    expect(readSceneBindings(next), next);
    expect(() => sceneEntityRemoval(next, 'b'), throwsStateError);
  });
  test('fixed descendants and stale confirmation prevent partial deletion', () {
    final scene = base(), patch = sceneEntityRemoval(base(), 'a')['patch'];
    scene['revision'] = 1;
    expect(() => applyScenePatch(scene, patch, byUser: true), throwsStateError);
    scene['entities'].firstWhere((e) => e['id'] == 'coat_a')['locked'] = true;
    final before = canonicalSceneValue(scene);
    expect(() => sceneEntityRemoval(scene, 'a'), throwsStateError);
    expect(canonicalSceneValue(scene), before);
  });
  test(
      'large confirmed user cascades work but do not enlarge model operation limit',
      () {
    final scene = base();
    scene['facts'] = List.generate(
        80,
        (i) => {
              'id': 'f$i',
              'entityId': 'coat_a',
              'slot': 's$i',
              'prompt': 'red'
            });
    final plan = sceneEntityRemoval(scene, 'coat_a');
    expect(plan['next']['facts'], isEmpty);
    expect(() => applyScenePatch(scene, plan['patch']), throwsStateError);
  });
}
